core/optaplanner-core-impl/src/main/java/org/optaplanner/core/api/domain/variable/VariableListener.java

```diff
@@ -86,6 +86,84 @@ default boolean requiresUniqueEntityEvents() {
      */
     void afterEntityRemoved(ScoreDirector<Solution_> scoreDirector, Entity_ entity);
 
+    /**
+     * @param scoreDirector never null
+     * @param fact never null
+     */
+    default void beforeProblemFactAdded(ScoreDirector<Solution_> scoreDirector, Object fact) {
+        // TODO remove default implementation in 9.0
+    }
+
+    /**
+     * For reasons of backwards compatibility,
+     * the default implementation will call {@link #resetWorkingSolution(ScoreDirector)}
+     * as any of the facts may have been used in the listener.
+     *
+     * @param scoreDirector never null
+     * @param fact never null
+     */
+    default void afterProblemFactAdded(ScoreDirector<Solution_> scoreDirector, Object fact) {
+        // TODO remove default implementation in 9.0
+        resetWorkingSolution(scoreDirector);
+    }
+
+    /**
+     * Call before either a problem fact changes or a planning entity changes without changing its planning variable.
+     *
+     * <p>
+     * <strong>Do not use this method when changing planning variable values on planning entities.</strong>
+     * Use {@link #beforeVariableChanged(ScoreDirector, Object)} instead.
+     * Failing to do so will result in score corruptions.
```

Comment:
This also looks more like a wording for the ProblemChangeDirector interface on which I decide whether to call this or not. Here on the other I would find useful any information that helps me decide whether I need to implement/override this method and what expectations I can have.